### ISSUE TITLE
Isolate imports and conditionally import libraries to make package platform-independent

### ISSUE THREAD

**Issue by sidrao2006** [Baseflow/flutter_cache_manager#325]
## 🏗 Enhancement Proposal

Currently `flutter_cache_manager` is platform-independent, yet according to pub.dev, the package doesn't support web. This can be solved using conditionally importing `dart:io` dependent libraries.
`path_provider`, `file` and `sqflite` are platform-dependent and must only be imported/exported in *_io.dart files which will then be conditionally imported/exported.

`file` can be replaced with `cross_file` (as mentioned in [#321 (comment)](https://github.com/Baseflow/flutter_cache_manager/pull/321#issuecomment-854064553))

`path_provider` and `sqflite` are rather easy to isolate since they are imported only in cache_info_repositories which are selectively imported in config_*.dart files. The only problem is that `cache_info_repositories/cache_info_repositories.dart` is exported. @renefloor what can we do about this?

### Pitch

This will increase pub points which affect discoverability. Also, partially platform-dependent code will be separated and be easier to maintain.

### Checklist

- [ ] #339 
- [ ] Store the config json data as a file on all platforms except the web, where web storage will be used to store this data
- [ ] Make the JSON implementation of cache info repository platform independent
- [ ] Migrate all instances to use the new JSON implementation
- [ ] If necessary, replace platform dependent dependencies with cross-platform implementations

**Comment by renefloor** [comment]
I think this is a good idea. I have been trying this already and did a lot of restructuring work. That's also the main reason the Config file is created to easily create a separate config for io and for web and not having to do the conditional import for all classes. However, I was mainly struggling with replacing the directory structure, but I made a start with that as well by creating the FileSystem class.

One of the reasons I did not continue working on it (besides I was kind of done with it) is that when pub.dev shows the package supports web devs might expect full caching support on web, but I've not found a good way to really support web besides just downloading the file.

**Comment by sidrao2006** [comment]
> web devs might expect full caching support on web, but I've not found a good way to really support web besides just downloading the file
@renefloor could you please elaborate on that?

**Comment by renefloor** [comment]
Yes, so at the moment this CacheManager doesn't cache anything on the web.
Replacing the database with cache information is relatively easy, so example [hive uses an indexedDB](https://github.com/hivedb/hive/blob/master/hive/lib/src/backend/js/backend_manager.dart). However, storing the files themselves is a bit more difficult. We already had a pretty long discussion with some small promise of how it might work here: #122.

**Comment by sidrao2006** [comment]
Hmm.. And what about Windows and Linux? We could use sqflite_ffi

**Comment by sidrao2006** [comment]
> However, storing the files themselves is a bit more difficult.

@renefloor what do you think of the [Cache API](https://web.dev/cache-api-quick-guide/)?

I think we should abstract platform-dependent implementations in the cache_info_repositories.. For web, I'd say the best option would be to use the `js`package and implement a new cache object provider.

**Comment by renefloor** [comment]
Windows and Linux are already compatible because those don't use sqflite, but they use a JSON file by default. I want to migrate all plaforms to using a JSON file by default and move sqflite to a separate package. I'm not sure if there is a way to do a conditional import based on platform to already fix this without the full migration. I think conditional imports are always io vs html.

**Comment by sidrao2006** [comment]
You read my mind! I was about to suggest migrating all platforms to use a json file.. Have you looked into [universal_io](https://pub.dev/packages/universal_io)?

You could simply merge the two config files into one which uses the JSON implementation and replace dart:io with universal_io

**Comment by sidrao2006** [comment]
As for path_provider, we can do a conditional import with a custom implementation for web

**Comment by sidrao2006** [comment]
@renefloor , would you be interested in rewriting the implementation for file system and using JSON files by default?

**Comment by renefloor** [comment]
That's already implemented, only the default has to switch and the sqflite version migrated. However you still have to get rid of dart:io for the MemoryFileSystem that is currently used for web.

**Comment by sidrao2006** [comment]
Hmm.. Right

But we would still have to get rid of the dart:io, path and path_provider imports in the JSON implementation.

So how about using Indexeddb to store files and localstorage to store the JSON data for the Web?

**Comment by sidrao2006** [comment]
So @renefloor, the main steps should be something like this -

- Implement file_system for the web using IndexedDB to store the files (more specifically, the file content as blobs)
- Store the config json data as a file on all platforms except the web, where web storage will be used to store this data
- Make the JSON implementation of cache info repository platform independent
- Migrate all instances to use the new JSON implementation
- If necessary, replace platform dependent dependencies with cross-platform implementations

If you are satisfied with the above mentioned changes, I'll go ahead and create separate issues and PRs to track each change and use this issue to track the overall migration to cross-platform implementations.

**Comment by renefloor** [comment]
@sidrao2006 this sounds like a good idea. I think we can also store the json file in the indexesDB?
1 thing I'm not sure about if whether we know how large the storage is and if we can do anything with that. This might also be a good improvement for the native implementation, but I think we might have to check for free storage and if that's to low delete the oldest file.

**Comment by sidrao2006** [comment]
@renefloor it'd be a bit more performant to store the json data from the json file using local storage but we could definitely use indexddb too.. 

`navigator.storage.estimate()` provides a way to check the total size of the indexeddb. I'm not sure about checking for free storage on the web but the above also provides a 'quota' field which lists the total storage allowed to be used by the website.
